Add command to list installed packages.

hiteshjasani opened the pull request:
A package manager should have a way to list installed packages, so I added one.

dom96 commented on src/nimble.nim:
```diff
@@ -663,6 +668,29 @@ proc list(options: TOptions) =
       echoPackageVersions(pkg)
     echo(" ")
 
+proc c_printf(frmt: cstring) {.importc: "printf", header: "<stdio.h>",
```
Use Nim's string formatting features please.

dom96 commented on src/nimble.nim:
```diff
@@ -663,6 +668,29 @@ proc list(options: TOptions) =
+proc listInstalled(options: TOptions) =
+  let pkgsDir = options.getNimbleDir() / "pkgs"
+  if not existsDir(pkgsDir):
+    raise newException(ENimble, "Nimble dir not installed.")
```
Change that to: ``raise newException(ENimble, "Could not find Nimble pkgs dir: " & pkgsDir)``

dom96 commented on src/nimble.nim:
```diff
@@ -55,6 +57,7 @@ Commands:
   search       [--ver] pkg/tag    Searches for a specified package. Search is
                                   performed by tag and by name.
   list         [--ver]            Lists all packages.
+  installed                       Lists all installed packages and versions.
```
I would prefer a flag for the ``list`` subcommand. ``--installed`` perhaps.

dom96 commented on src/nimble.nim:
```diff
@@ -663,6 +668,29 @@ proc list(options: TOptions) =
+  if not existsDir(pkgsDir):
+    raise newException(ENimble, "Nimble dir not installed.")
+  var h = initTable[string, seq[string]]()
+  for kind, path in walkDir(pkgsDir):
```
You should use ``packageinfo.getInstalledPkgs`` here.

hiteshjasani commented:
A couple of thoughts:
* From what I can tell, Nim's string formatting features are nice but do not match up with the features and terseness of `printf` quite yet.  I have some ideas about how to improve the string formatting, but that would need to happen outside of nimble.  The end result is that I've removed the pretty printing of package names and versions.
* I moved the option to under the `list` command using either `-i` or `--installed`.  I wish the option parser in nimble were more advanced to avoid using up the `-i` option for only the list command.  But it's not worth reimplementing the option parser just to add this simple command.  On the flip side, being able to list installed packages should be a very short command, which `nimble list -i` accomplishes so I'm happy with the result.

dom96 commented:
Brilliant thank you!

Once something else starts using ``-i`` we can improve the situation.
